### PR TITLE
FP8 params support for megatron-fsdp (MXFP8/Blockwise)

### PR DESCRIPTION
**Pull request by kunlunl**
# What does this PR do ?

dev MR: https://github.com/NVIDIA/Megatron-LM/pull/2086

Major changes:

Make Megatron-FSDP’s AG pipeline support using different data parallel buffers, because MXFP8 has different quantization direction in forward and backward passes.
Decouple the FP8-related logic from the main workflow and provide a unified abstraction to 1) operate the raw data storage of different recipes; 2) Create or discard transpose cache for different recipes.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by Skylion007** [comment]
Is FP8 activations/grad support on Hopper FSDP with block wise support on the roadmap as well? O-o

**Comment by kunlunl** [comment]
> Is FP8 activations/grad support on Hopper FSDP with block wise support on the roadmap as well? O-o

@shjwudp

**Comment by cspades** [comment]
See comment on Dev PR: https://github.com/NVIDIA/Megatron-LM/pull/2086#issuecomment-3661679668

Can we add a few simple unit tests?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/mixed_precision.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    if not isinstance(TE_VERSION, PkgVersion):
+        return False
+
+    if check_equality:
+        return TE_VERSION >= PkgVersion(vers)
+    else:
+        return TE_VERSION > PkgVersion(vers)
+
+
+def is_float8tensor(tensor: torch.Tensor) -> bool:
```
Comment (Skylion007) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/mixed_precision.py:157]:
```suggestion
def is_float8tensor(tensor: torch.Tensor) -> TypeGuard[FP8_TENSOR_CLASS]:

**Comment by shjwudp** [comment]
> Is FP8 activations/grad support on Hopper FSDP with block wise support on the roadmap as well? O-o

Hi @Skylion007 , 

Do you mean FP8 activations + param support on Hopper? We has already been merged into the dev branch ([PR-2086](https://github.com/NVIDIA/Megatron-LM/pull/2086)), and we'll look into merging it into the main branch when we have time.

**Comment by cspades** [comment]
/ok to test [42a6bdc](https://github.com/NVIDIA/Megatron-LM/pull/2239/commits/42a6bdcca0e812204230090ca54f489321fefe2c)

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 2]
```diff
@@ -564,12 +571,15 @@ def _post_backward(module, *unused):
...
                     # Deallocate the module parameters after the backward pass,
                     # because we have our data-parallel gradients computed.
-                    release_module_parameters(module)
+                    release_module_parameters(module, bwd=True)
                     module._training_state = TrainingState.IDLE
                 param_list = list(module.parameters())
             else:
                 param_list = list(module.parameters(recurse=False))
 
+            if self.enable_fine_grained_param_gather_hook:
```
Comment (BoxiangW) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:580]: Can we combine this if condition with the above?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 3]
```diff
@@ -621,6 +631,9 @@ def _pre_forward_param_unshard(
                 # to allocate as little memory as possible for this forward pass.
                 param_list = list(module.parameters(recurse=False))
 
+            if self.enable_fine_grained_param_gather_hook:
```
Comment (BoxiangW) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:634]: Same here

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 4]
```diff
@@ -729,11 +742,19 @@ def _pre_backward(module: nn.Module, *unused):
...
             if isinstance(module, tuple(fsdp_unit_modules)):
-                # All-gather / unshard the module parameters before the backward pass.
-                self.all_gather_and_wait_parameters_ready(
-                    list(module.parameters()), prefetch_order=PrefetchOrder.BACKWARD_PASS_ORDER
-                )
+                param_list = list(module.parameters())
+            else:
+                param_list = list(module.parameters(recurse=False))
+
+            if self.enable_fine_grained_param_gather_hook:
```
Comment (BoxiangW) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:751]: same here

**Review by cspades — APPROVED** [review]
Finally found time to prototype this backend implementation in `fully_shard`, I'm generally happy with this PR. I'll submit a follow-up PR directly to `main` that exposes FP8 parameter support to `fully_shard` and a unit test for it as well.

@shjwudp I do have a comment beyond the scope of this PR though, pertaining to this `fp8_model_init`: https://github.com/NVIDIA/Megatron-LM/blob/dev/megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py#L3738

We should move this to `mcore_fsdp_adapter.py` during `MegatronFSDP.__init__` so both in Megatron-LM and native Torch, we can have the same initialization pattern:
```
# Construct toy model.
with te.pytorch.quantized_model_init(
    enabled=True,
    recipe=fp8_recipe,
    # Needed for FP8 parameters with Megatron-FSDP.
    preserve_high_precision_init_val=True,
):
    toy_model = ToyTETransformer(
        model_dim=DIM_SIZE,
        num_heads=2,
        num_layers=NUM_LAYERS,
        output_dim=DIM_SIZE,
        device="meta"
    )

# Fully-shard the model.
# NOTE: We do NOT need the quantized_model_init context manager for Megatron-FSDP,
# because it has already setup the correct state during the root module FP8 init, I believe?
mfsdp_model = fully_shard_model(
    module=toy_model,
    fsdp_unit_modules=[te.pytorch.TransformerLayer, te.pytorch.Linear],
    zero_dp_strategy=3,
    init_model_with_meta_device=True,
)
```
Agree? I believe this also means we do not need an `fp8_param_gather` argument either!

**Comment by cspades** [comment]
/ok to test [cba67e3](https://github.com/NVIDIA/Megatron-LM/pull/2239/commits/cba67e3d551586fcaa52f97abca023f35bb4c759)

**Comment by cspades** [comment]
API compatibility check error is expected, just like in the DEV PR, with the same violations.

**Review by kvareddy — APPROVED** [review]

**Review by BoxiangW — APPROVED** [review]
